### PR TITLE
Sp/tuna machine typehint

### PR DESCRIPTION
**Pull request by technicalgrp89**
machine.py - code is made compatible with mypy stds. Varables/functions/input/output/retrun parameters data types are defined.

### CODE REVIEW

File: tuna/machine.py [hunk 1]
```diff
@@ -283,31 +306,48 @@ def get_properties(self):
...
+    fout_2: BufferedWriter
+
+    t_filename: Union[str, 'os.PathLike[Any]']
     if is_temp:
       assert filename is None
-      _, filename = tempfile.mkstemp()
-    else:
-      assert filename is not None
+      _, t_filename = tempfile.mkstemp()
+      return t_filename
```
Comment (alexandraBara) [inline tuna/machine.py]: if this returns here, the contents never gets written to the json, hence and empty json file and FIN will complain the json is not as expected. Please address by removing the return statement here

File: tuna/machine.py [hunk 2]
```diff
@@ -283,31 +306,55 @@ def get_properties(self):
...
+  def write_file(self, contents: bytes, filename: Union[str, bytes, 'os.PathLike[str]',\
+  None] = None, is_temp: bool = False) -> Union[str, bytes, 'os.PathLike[str]', None]:
     """
     Write a file to this machine containing contents
     """
+    cnx: Connection
+    ftp: Optional[paramiko.sftp_client.SFTPClient] = None
+    fout: SFTPFile
+    fout_2: IO[Any]
+    t_file: bool = False
```
Comment (alexandraBara) [inline tuna/machine.py]: this is not needed, 1 variable is enough to hold the temp file. You should just return filename

File: tuna/machine.py [hunk 3]
```diff
@@ -283,31 +306,55 @@ def get_properties(self):
...
-      ftp = cnx.ssh.open_sftp()
-      with ftp.open(filename, 'wb') as fout:
-        fout.write(contents)
-        fout.flush()
+      if cnx is not None and cnx.ssh is not None:
+        ftp = cnx.ssh.open_sftp()
+      if ftp is not None:
+        with ftp.open(filename, 'wb') as fout:  #type: ignore
+          fout.write(contents)
+          fout.flush()
```
Comment (alexandraBara) [inline tuna/machine.py]: this is why you get the error, this code is not reached, you have changed the way the code works here, please undo

**Review by alexandraBara — CHANGES_REQUESTED** [review]

File: tuna/machine.py [hunk 4]
```diff
@@ -283,31 +307,46 @@ def get_properties(self):
...
+
+    #t_filename: Union[str, 'os.PathLike[str]']
     if is_temp:
       assert filename is None
       _, filename = tempfile.mkstemp()
     else:
       assert filename is not None
-    if self.local_machine:  # pylint: disable=no-member ; false alarm
+
+    if self.local_machine and filename is not None:  # pylint: disable=no-member ; false alarm
```
Comment (alexandraBara) [inline tuna/machine.py]: i dont think the 'filename is not none' part is required here. The filename is established beforehand or passed in as not none.

File: tuna/machine.py [hunk 5]
```diff
@@ -283,31 +307,46 @@ def get_properties(self):
...
-  def write_file(self, contents, filename=None, is_temp=False):
+  def write_file(self, contents: bytes, filename: Union[str, bytes, 'os.PathLike[str]',\
+  None] = None, is_temp: bool = False) -> Union[str, bytes, 'os.PathLike[str]', None]:
     """
     Write a file to this machine containing contents
     """
+    cnx: Connection
+    ftp: Optional[paramiko.sftp_client.SFTPClient] = None
+
+    #t_filename: Union[str, 'os.PathLike[str]']
```
Comment (alexandraBara) [inline tuna/machine.py]: remove

File: tuna/machine.py [hunk 6]
```diff
@@ -319,21 +358,28 @@ def read_file(self, filename, byteread=False):
       ftp.getfo(filename, content_io)
       ret = content_io.getvalue()
       if not byteread:
-        ret = ret.decode()
+        ret_str = ret.decode()
+        return ret_str
```
Comment (alexandraBara) [inline tuna/machine.py]: can you set ret as Any[str, bytes] and return ret, thus removing the extra ret_str and leave the code as it was to just return ret.

File: tuna/machine.py [hunk 7]
```diff
@@ -377,25 +432,15 @@ def get_gpu_clock(self, gpu_num=0):
...
-      mclk_cmd = f'{ROCMSMI} --setmclk {self.mclk}'
-      cnx.exec_command(mclk_cmd)
-
-  def restart_server(self, wait=True):
+  def restart_server(self, wait: bool = True) -> bool:
     """restart the machine"""
     logger = self.get_logger()
 
     logger.warning('Sending remote reboot command')
     if self.ipmi_ip is not None and not self.ipmi_inaccessible:
```
Comment (alexandraBara) [inline tuna/machine.py]:
```suggestion
    if self.ipmi_ip is not None and not self.ipmi_inaccessible and self.mmi:
```

File: tuna/machine.py [hunk 8]
```diff
@@ -377,25 +432,15 @@ def get_gpu_clock(self, gpu_num=0):
...
       logger.info('Using IPMI to reboot machine')
-      self.mmi.restart_server()
+      if self.mmi is not None:
```
Comment (alexandraBara) [inline tuna/machine.py]: remove this condition and expand the above one like i suggested

**Review by alexandraBara — APPROVED** [review]

**Comment by alexandraBara** [comment]
sorry clicked 'Approved' by mistake, i meant to do 'Request changes' alongside my comments :)

File: tuna/machine.py [hunk 9]
```diff
@@ -186,22 +207,22 @@ def get_gpu(self, idx):
...
 
-  def get_cpus(self):
-    """return cpu details array"""
-    if not self.cpus:
-      self.get_properties()
-
-    return self.cpus
-
-  def parse_agents(self):  #pylint: disable=too-many-locals
+  def parse_agents(self) -> dict:  #pylint: disable=too-many-locals
```
Comment (cderb) [inline tuna/machine.py:209]: shouldn't disable pylint

File: tuna/machine.py [hunk 10]
```diff
@@ -83,63 +93,73 @@ def __init__(self, **kwargs):
...
-    self.log_list = {}
-    self.num_cpus = 0
+    self.cnx_list: dict = {}
+    self.log_list: dict = {}
+    self.num_cpus: int = 0
+    self.avail_gpus: List[Dict[int, str]]
+    self.sclk: int
+    self.mclk: int
+    self.gpus: List[Dict[Any, Any]]
+    self.cpus: List[dict]
```
Comment (cderb) [inline tuna/machine.py]: inconsistent with get_properties return type.

File: tuna/machine.py [hunk 11]
```diff
@@ -283,31 +307,46 @@ def get_properties(self):
 
     return self.cpus, self.gpus
 
-  def write_file(self, contents, filename=None, is_temp=False):
+  def write_file(self, contents: bytes, filename: Union[str, bytes, 'os.PathLike[str]',\
```
Comment (cderb) [inline tuna/machine.py]: the filename as an input to this function should only be a string

File: tuna/machine.py [hunk 12]
```diff
@@ -347,14 +393,23 @@ def exec_command(self, command, timeout=LOG_TIMEOUT):
...
+    stdout = Optional[IO[str]]
+    gpu_clk_cmd: str = f'{ROCMSMI} -c'
     _, stdout, _ = self.connect().exec_command(gpu_clk_cmd)
 
-    base_idx = None
-    gpu_clk = []
+    base_idx: Union[int, None] = None
+    gpu_clk: List[Dict[str, int]] = []
+    gpu_idx: int = 0
+    level: Union[int, str]
```
Comment (cderb) [inline tuna/machine.py]: appears to be int only

File: tuna/machine.py [hunk 13]
```diff
@@ -347,14 +393,23 @@ def exec_command(self, command, timeout=LOG_TIMEOUT):
...
+    sclk: Union[SupportsIndex, slice]
+    mclk: Union[SupportsIndex, slice]
```
Comment (cderb) [inline tuna/machine.py]: sclk and mclk are clock levels, level is an int.

File: tuna/machine.py [hunk 14]
```diff
@@ -408,12 +453,14 @@ def restart_server(self, wait=True):
...
+  def chk_gpu_status(self, gpu_id: int) -> bool:
     """check gpu status, can clinfo find the device"""
-    logger = self.get_logger()
+    line: str
+    stdout: ChannelFile
+    logger: logging.Logger = self.get_logger()
     cnx = self.connect()
 
-    if gpu_id not in self.avail_gpus:
+    if gpu_id not in self.avail_gpus:  #type: ignore
```
Comment (cderb) [inline tuna/machine.py]: type ignore?

File: tuna/machine.py (hunk 5 above)
Comment (technicalgrp89) [reply tuna/machine.py]: ok

File: tuna/machine.py (hunk 11 above)
Comment (technicalgrp89) [reply tuna/machine.py]: ok. filename input changed to str type

File: tuna/machine.py (hunk 4 above)
Comment (technicalgrp89) [reply tuna/machine.py]: ok

File: tuna/machine.py (hunk 9 above)
Comment (alexandraBara) [reply tuna/machine.py:209]: this should be addressed in a different PR

File: tuna/machine.py (hunk 6 above)
Comment (technicalgrp89) [reply tuna/machine.py]: I used Union[str, Bytes].

File: tuna/machine.py (hunk 7 above)
Comment (technicalgrp89) [reply tuna/machine.py:440]: check added for self.mmi

File: tuna/machine.py (hunk 8 above)
Comment (technicalgrp89) [reply tuna/machine.py]: check is removed and added in line 440

File: tuna/machine.py (hunk 7 above)
Comment (technicalgrp89) [reply tuna/machine.py]: self.mmi is not None added

File: tuna/machine.py (hunk 9 above)
Comment (technicalgrp89) [reply tuna/machine.py:209]: I havnot edited nor changed this condition it is retained as it is.

File: tuna/machine.py (hunk 10 above)
Comment (technicalgrp89) [reply tuna/machine.py]: corrected.  def get_properties(self): -> Tuple[List[Dict], List[Dict[Any, Any]]]

File: tuna/machine.py (hunk 12 above)
Comment (technicalgrp89) [reply tuna/machine.py]: Level type changed to int

File: tuna/machine.py (hunk 13 above)
Comment (technicalgrp89) [reply tuna/machine.py]: variable type correctd.

File: tuna/machine.py (hunk 14 above)
Comment (technicalgrp89) [reply tuna/machine.py]: removed.

**Review by alexandraBara — APPROVED** [review]

File: tuna/machine.py [hunk 15]
```diff
@@ -43,37 +50,39 @@
...
+  __tablename__: str = "machine"
+  hostname: str = Column(Text, nullable=False)
+  port: int = Column(INTEGER(11, unsigned=True),
+                     nullable=False,
+                     server_default="22")
+  local_ip: str = Column(Text, nullable=True)
+  local_port: int = Column(INTEGER, server_default="22")
+  user: str = Column(Text, nullable=False)
+  password: str = Column(Text, nullable=False)
+  avail_gpus: List[Dict[int, str]] = Column(Text, nullable=False)
```
Comment (cderb) [inline tuna/machine.py]: This is a List[int], see line 119.

File: tuna/machine.py [hunk 16]
```diff
@@ -83,63 +92,73 @@ def __init__(self, **kwargs):
     self.__dict__.update(
         (key, value) for key, value in kwargs.items() if key in allowed_keys)
 
-    self.cnx_list = {}
-    self.log_list = {}
-    self.num_cpus = 0
+    self.cnx_list: dict = {}
+    self.log_list: dict = {}
+    self.num_cpus: int = 0
+    self.avail_gpus: List[Dict[int, str]]
```
Comment (cderb) [inline tuna/machine.py]: wrong type

File: tuna/machine.py [hunk 17]
```diff
@@ -83,63 +92,73 @@ def __init__(self, **kwargs):
...
       self.logger = setup_logger('Machine_{self.id}')
       self.mmi = MachineManagementInterface(self.ipmi_ip, self.ipmi_port,
                                             self.ipmi_user, self.ipmi_password)
 
       if not self.avail_gpus is None:
-        self.avail_gpus = [int(val) for val in self.avail_gpus.split(',')]
+        self.avail_gpus = [
+            int(val) for val in self.avail_gpus.split(',')  #type: ignore
+        ]  #type: ignore
         self.num_gpus = len(self.avail_gpus)
```
Comment (cderb) [inline tuna/machine.py:136]: avail_gpus could come in through kwargs as a str. So it's ok to ignore that type complication here. This input parsing should probably be moved outside of this constructor in a separate PR.

File: tuna/machine.py [hunk 18]
```diff
@@ -83,63 +92,73 @@ def __init__(self, **kwargs):
...
-      self.id = 0  # pylint: disable=invalid-name
+
+      self.id: int = 0  # pylint: disable=invalid-name
       self.get_properties()
 
       if self.gpus:
         self.arch = self.gpus[0]['arch']
         self.num_cu = self.gpus[0]['num_cu']
-        self.avail_gpus = list(range(len(self.gpus)))
+        self.avail_gpus = list(range(len(self.gpus)))  #type: ignore
```
Comment (cderb) [inline tuna/machine.py]: no need to ignore type here

File: tuna/machine.py [hunk 19]
```diff
@@ -161,23 +180,24 @@ def connect(self, abort=None):
...
+  def get_num_cpus(self) -> int:
     """return number of available cpus"""
+    stdout: TextIO
     _, stdout, _ = self.connect().exec_command('nproc')
     self.num_cpus = int(stdout.readline())
 
     return self.num_cpus
 
-  def get_avail_gpus(self):
+  def get_avail_gpus(self) -> List[Dict[int, str]]:
```
Comment (cderb) [inline tuna/machine.py]: wrong type

File: tuna/machine.py [hunk 20]
```diff
@@ -161,23 +180,24 @@ def connect(self, abort=None):
...
     """return list of available gpus"""
     if not self.avail_gpus:
       if not self.gpus:
         self.get_properties()
-        self.avail_gpus = range(len(self.gpus))
+        self.avail_gpus = range(len(self.gpus))  #type: ignore
```
Comment (cderb) [inline tuna/machine.py]: don't ignore type.

**Review by cderb — CHANGES_REQUESTED** [review]

File: tuna/machine.py (hunk 15 above)
Comment (technicalgrp89) [reply tuna/machine.py]: changed.

File: tuna/machine.py (hunk 16 above)
Comment (technicalgrp89) [reply tuna/machine.py]: changed

File: tuna/machine.py (hunk 17 above)
Comment (technicalgrp89) [reply tuna/machine.py:136]: Ok. separate PR is ideal.

File: tuna/machine.py (hunk 18 above)
Comment (technicalgrp89) [reply tuna/machine.py]: removed

File: tuna/machine.py (hunk 20 above)
Comment (technicalgrp89) [reply tuna/machine.py]: error: Incompatible types in assignment (expression has type "range", variable has type "List[int]") 
used list() o enable the same

File: tuna/machine.py (hunk 19 above)
Comment (technicalgrp89) [reply tuna/machine.py]: changed to List[int]

File: tuna/machine.py (hunk 16 above)
Comment (cderb) [reply tuna/machine.py]: List[int]

File: tuna/machine.py (hunk 10 above)
Comment (cderb) [reply tuna/machine.py]: Dict[Any, Any] = dict, please use most succinct notation.
The keys for these dicts are always str, so Dict[str, Any] would be the most specific.

File: tuna/machine.py [hunk 21]
```diff
@@ -161,23 +180,24 @@ def connect(self, abort=None):
...
     if not self.avail_gpus:
       if not self.gpus:
         self.get_properties()
-        self.avail_gpus = range(len(self.gpus))
+        self.avail_gpus = list(range(len(self.gpus)))
         self.num_gpus = len(self.avail_gpus)
     return self.avail_gpus
 
-  def get_gpu(self, idx):
+  def get_gpu(self, idx: int) -> Optional[Dict[int, str]]:
```
Comment (cderb) [inline tuna/machine.py]: Optional[Dict[str, Any]]

**Review by cderb — CHANGES_REQUESTED** [review]

File: tuna/machine.py (hunk 21 above)
Comment (technicalgrp89) [reply tuna/machine.py]: Dict definitions are changed

File: tuna/machine.py [hunk 22]
```diff
@@ -322,18 +358,24 @@ def read_file(self, filename, byteread=False):
...
 
-  def make_temp_file(self):
+  def make_temp_file(self) -> Union[str, Text]:
     """
     Make an empty temp file on this machine
     """
     return self.write_file(b'', is_temp=True)
 
-  def exec_command(self, command, timeout=LOG_TIMEOUT):
+  def exec_command(self, command: str, timeout: int = LOG_TIMEOUT) -> Tuple[int, str,\
```
Comment (cderb) [inline tuna/machine.py]: Return value is the same as Connection.exec_command
Which is Tuple[int, StringIO, StringIO]
Remove line 387+388 because err returned from call is a StringIO and not a ChannelFile.

**Review by cderb — CHANGES_REQUESTED** [review]

File: tuna/machine.py (hunk 22 above)
Comment (technicalgrp89) [reply tuna/machine.py]: 387,388 removed. 
Line number 395: function get_gpu_clock, if stdout is null case added, as return False.

Comment (cderb) [reply tuna/machine.py]: out is not a str.

File: tuna/machine.py [hunk 23]
```diff
@@ -408,9 +448,11 @@ def restart_server(self, wait=True):
       return True
     return False
 
-  def chk_gpu_status(self, gpu_id):
+  def chk_gpu_status(self, gpu_id: int) -> bool:
     """check gpu status, can clinfo find the device"""
-    logger = self.get_logger()
+    line: str
+    stdout: ChannelFile
```
Comment (cderb) [inline tuna/machine.py]: cnx.exec_command no longer returns stdout as a channel file, stdout is a StringIO. Remove lines that treat it as a ChannelFile.

File: tuna/machine.py [hunk 24]
```diff
@@ -442,14 +484,19 @@ def chk_gpu_status(self, gpu_id):
...
-  def getusedspace(self):
+  def getusedspace(self) -> float:
     """examine used space on disk"""
-    logger = self.get_logger()
+    logger: logging.Logger = self.get_logger()
     logger.info("Getting free space")
+    lst: list
+    per: float
+    output: bytes
+    ssh_stdout: ChannelFile
```
Comment (cderb) [inline tuna/machine.py]: ssh_stdout is StringIO.

**Review by cderb — CHANGES_REQUESTED** [review]

File: tuna/machine.py (hunk 23 above)
Comment (technicalgrp89) [reply tuna/machine.py]: if hasattr(stdout, 'channel'):
      stdout.channel.settimeout(30) , this ;ine leads to machine.py:467: error: "StringIO" has no attribute "channel"

Comment (technicalgrp89) [reply tuna/machine.py]: an we remove this check/. or we can do it on err output, as done in different classes.  if err is not None and hasattr(err, 'channel'): as in worker interface. Let me know

File: tuna/machine.py (hunk 24 above)
Comment (technicalgrp89) [reply tuna/machine.py]: changed

File: tuna/machine.py (hunk 22 above)
Comment (technicalgrp89) [reply tuna/machine.py]: changed

File: tuna/machine.py (hunk 23 above)
Comment (cderb) [reply tuna/machine.py]: remove this check, as StringIO does not have this attribute.

**Review by cderb — CHANGES_REQUESTED** [review]
Incorrect types still present around cnx.exec_command calls

File: tuna/machine.py (hunk 23 above)
Comment (technicalgrp89) [reply tuna/machine.py]: check removed.

File: tuna/machine.py [hunk 25]
```diff
@@ -322,18 +358,24 @@ def read_file(self, filename, byteread=False):
...
+  def exec_command(self, command: str, timeout: int = LOG_TIMEOUT) -> Tuple[int, StringIO,\
+  StringIO]:
     """
     Execute a command on this machine
     - through docker if on a remote machine
     - no docker on local machine
     """
+    cnx: Connection
+    ret_code: int
+    out: StringIO = StringIO()
```
Comment (cderb) [inline tuna/machine.py:376]: out: StringIO
is sufficient

File: tuna/machine.py [hunk 26]
```diff
@@ -342,19 +384,27 @@ def exec_command(self, command, timeout=LOG_TIMEOUT):
...
-      err.channel.settimeout(LOG_TIMEOUT)
 
     return ret_code, out, err
 
-  def get_gpu_clock(self, gpu_num=0):
+  def get_gpu_clock(self, gpu_num: int = 0) -> Union[Tuple[int, int], bool]:
     """query gpu clock levels with rocm-smi"""
-
-    gpu_clk_cmd = f'{ROCMSMI} -c'
+    stdout: StringIO = StringIO()
```
Comment (cderb) [inline tuna/machine.py:392]: stdout: StringIO
is sufficient

File: tuna/machine.py [hunk 27]
```diff
@@ -408,9 +447,11 @@ def restart_server(self, wait=True):
       return True
     return False
 
-  def chk_gpu_status(self, gpu_id):
+  def chk_gpu_status(self, gpu_id: int) -> bool:
     """check gpu status, can clinfo find the device"""
-    logger = self.get_logger()
+    line: str
+    stdout: StringIO = StringIO()
```
Comment (cderb) [inline tuna/machine.py:453]: stdout: StringIO
is sufficient

File: tuna/machine.py [hunk 28]
```diff
@@ -442,14 +481,19 @@ def chk_gpu_status(self, gpu_id):
...
-  def getusedspace(self):
+  def getusedspace(self) -> float:
     """examine used space on disk"""
-    logger = self.get_logger()
+    logger: logging.Logger = self.get_logger()
     logger.info("Getting free space")
+    lst: list
+    per: float
+    output: str
+    ssh_stdout: StringIO = StringIO()
```
Comment (cderb) [inline tuna/machine.py:491]: ssh_stdout: StringIO
is sufficient

**Review by cderb — APPROVED** [review]
Technically correct now. Would request reducing commented verbose lines.
